Fixes #25998 - pulp3 file creation

jlsherrill opened the pull request:
Left to do:

* [x] Tests
* [x] add foreign keys and indexes
* [x] record walkthrough video of design

Highly Recommend watching: https://www.youtube.com/playlist?list=PL59WSIBtIpv7tbjnsEAWNcdiOKAmjjgqE

jturel commented on app/models/katello/concerns/smart_proxy_extensions.rb:
```diff
@@ -126,6 +126,14 @@ def pulp3_api
 
+      def backend_service_type(repository)
+        if pulp3_support?(repository)
+          Actions::Pulp3::Abstract::BACKEND_SERVICE_TYPE
```
Need to think on it more but I wonder if this constant should come from somewhere other than an action class. Probably a nitpick either way

jlsherrill replied:
Yeah, i couldn't think of a better place, right now actions are the only ones utilizing it

parthaa replied:
would it be better if this were a method instead of a constant?

jlsherrill commented on app/services/katello/pulp3/repository.rb:
```diff
@@ -0,0 +1,152 @@
+          }
+        else
+          {}
+        end
```
One thing to ponder, should we pull distribution/remote/publisher out into their own classes?   I'm leaning that way, but might make sense in a future pr rather than this one

jlsherrill commented:
[test katello]

jlsherrill commented:
Pushed a couple of changes:
* Updated distribution creation/update to be async (this introduces new complexity around saving the distribution_href)
* moved initial repository version creation to metadata generation, since it is async

sjha4 commented:
Testing with this, everyhting works well except I noticed that 2 distributors(http and https) were created for every repo in pulp. However the distribution references were not created in katello. I did see an error in the metadata generate task in Actions::Pulp3::Repository::RefreshDistribution which errored with the following and then paused. Upon skip and resume, the Actions::Pulp3::Repository::SaveDistributionReferences task also got skipped.
```

NoMethodError

undefined method `[]' for #<Zest::Distribution:0x00007fa810545500>

---
- "/home/vagrant/katello/app/lib/katello/errors.rb:103:in `from_task'"
- "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:109:in `block
  in external_task='"
- "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:108:in `each'"
- "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:108:in `external_task='"
- "/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.2.2/lib/dynflow/action/polling.rb:83:in
  `initiate_external_action'"

```

jlsherrill commented:
What version of zest are you using? Can you see if you're on 0.0.4?

On Thu, Mar 21, 2019, 11:04 PM Samir Jha <notifications@github.com> wrote:

> Testing with this, everyhting works well except I noticed that 2
> distributors(http and https) were created for every repo in pulp. However
> the distribution references were not created in katello. I did see an error
> in the metadata generate task in
> Actions::Pulp3::Repository::RefreshDistribution which errored with the
> following and then paused. Upon skip and resume, the
> Actions::Pulp3::Repository::SaveDistributionReferences task also got
> skipped.
>
>
> NoMethodError
>
> undefined method `[]' for #<Zest::Distribution:0x00007fa810545500>
>
> ---
> - "/home/vagrant/katello/app/lib/katello/errors.rb:103:in `from_task'"
> - "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:109:in `block
>   in external_task='"
> - "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:108:in `each'"
> - "/home/vagrant/katello/app/lib/actions/pulp3/abstract_async_task.rb:108:in `external_task='"
> - "/home/vagrant/foreman/.vendor/ruby/2.5.0/gems/dynflow-1.2.2/lib/dynflow/action/polling.rb:83:in
>   `initiate_external_action'"
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Katello/katello/pull/7992#issuecomment-475478058>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYHRW0Ydvk7ExuKGDHsXxqw8Z4eWwElks5vZEg4gaJpZM4bZTS2>
> .
>

sjha4 commented:
> What version of zest are you using? Can you see if you're on 0.0.4?

Ofcourse I was on an older version.. Will test again.. Me--

sjha4 commented:
Works perfectly with 0.0.4! Distributions created and references created.. :tada: :tada:

sjha4 commented:
Couple questions:
1. What are the issues with a single publisher per content type?
2. Should we create remote when creating repo or can we defer that to the first sync?

jlsherrill commented:
>     What are the issues with a single publisher per content type?

The problem with that is that task locking will prevent multiple publishes on the same publisher from being run, so you'd only be able to publish a single yum repo at a time (for example).

jlsherrill commented:
[test katello]

jlsherrill commented:
@sjha4 and @jjeffers ya'll have played with this a bit now and @parthaa you've looked at it a bit, any objection to merging?  @parthaa want to look a bit more?

sjha4 commented:
This has been working great for me! ACK :+1:

jlsherrill commented:
@sjha4 mind doing an official review approval?  @parthaa thoughts?

parthaa commented:
ACK. Code wise looks ok to me.

tbrisker commented on katello.gemspec:
```diff
@@ -42,7 +42,7 @@ Gem::Specification.new do |gem|
   gem.add_dependency "runcible", ">= 2.11.0", "< 3.0.0"
   gem.add_dependency "anemone"
-
+  gem.add_dependency "zest"
```
This is missing packaging, now katello builds are failing

tbrisker replied:
https://github.com/theforeman/foreman-packaging/pull/3592
